lib/contract.ts

```diff
@@ -0,0 +1,837 @@
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import * as cryptojs from 'crypto-js'
+import * as crypto from 'crypto'
+import bs58 from './bs58'
+import fs from 'fs'
+import { promises as fsPromises } from 'fs'
+import { CliqueClient } from './clique'
+import * as api from '../api/api-alephium'
+import { Signer } from './signer'
+
+export abstract class Common {
+  readonly fileName: string
+  readonly sourceCodeSha256: string
+  readonly bytecode: string
+  readonly codeHash: string
+  readonly functions: api.Function[]
+
+  static readonly importRegex = new RegExp('^import "[a-z][a-z_0-9]*.ral"', 'm')
+  static readonly contractRegex = new RegExp('^TxContract [A-Z][a-zA-Z0-9]*\\(', 'm')
+  static readonly scriptRegex = new RegExp('^TxScript [A-Z][a-zA-Z0-9]* \\{', 'm')
+  static readonly variableRegex = new RegExp('\\{\\{\\s+[a-z][a-zA-Z0-9]*\\s+\\}\\}', 'g')
+
+  constructor(
+    fileName: string,
+    sourceCodeSha256: string,
+    bytecode: string,
+    codeHash: string,
+    functions: api.Function[]
+  ) {
+    this.fileName = fileName
+    this.sourceCodeSha256 = sourceCodeSha256
+    this.bytecode = bytecode
+    this.codeHash = codeHash
+    this.functions = functions
+  }
+
+  protected static _contractPath(fileName: string): string {
+    return `./contracts/${fileName}`
+  }
+
+  protected static _artifactPath(fileName: string): string {
+    return `./artifacts/${fileName}.json`
+  }
+
+  protected static _artifactsFolder(): string {
+    return './artifacts/'
+  }
+
+  static async handleImports(contractStr: string, importsCache: string[]): Promise<string> {
+    const localImportsCache: string[] = []
+    let result = contractStr.replace(Common.importRegex, (match) => {
+      localImportsCache.push(match)
+      return ''
+    })
+    for (const myImport of localImportsCache) {
+      const fileName = myImport.slice(8, -1)
+      if (!importsCache.includes(fileName)) {
+        importsCache.push(fileName)
+        const importContractStr = await Common._loadContractStr(fileName, importsCache, (code) =>
+          Contract.checkCodeType(fileName, code)
+        )
+        result = result.concat('\n', importContractStr)
+      }
+    }
+    return result
+  }
+
+  protected static _replaceVariables(contractStr: string, variables?: any): string {
+    if (variables) {
+      return contractStr.replace(Common.variableRegex, (match) => {
+        const variableName = match.split(/(\s+)/)[2]
+        if (variableName in variables) {
+          return variables[variableName].toString()
+        } else {
+          throw new Error(`The value of variable ${variableName} is not provided`)
+        }
+      })
+    } else {
+      return contractStr
+    }
```

Comment:
```suggestion
    if (variables === undefined) return contractStr
    return contractStr.replace(Common.variableRegex, (match) => {
        const variableName = match.split(/(\s+)/)[2]
        if (variableName in variables) {
            return variables[variableName].toString()
        } else {
            throw new Error(`The value of variable ${variableName} is not provided`)
        }
    })
```

> And this is labeled as critical: https://rules.sonarsource.com/typescript/RSPEC-126

This is labeled as critical.

This is not related to your code. Where are you using **else if**? I think you are not understanding the rule.

I think the code could be further improved but I have not looked into the usage of the function. I'll stick to my conservative suggestion.

Reply:
I still find the use of `else` there useless.

Reply:
Ah yep, inside the callback it should be removed too. Updating.

Reply:
Could also do `!!variables` if we want to check for other falsy values.

Reply:
I was saying that provides insights, not exactly the case. If we insist an else clause for `if ... else if ..`, why not using else clause for consistency?

I don't find typescript coding guide for `if-else` statement, but google's c/c++ coding style is explicit about this: https://google.github.io/styleguide/cppguide.html#Conditionals

TBH, I don't see a reason to discourage this coding style.

Reply:
You have code elsewhere @nop33 pointed out where you omit the `else`. Are you arguing against yourself?

Reply:
It's the style of `if (unhappyCase)` as I mentioned on Slack. And THANK YOU for bringing this up again here 👍

Reply:
Is it really recommended to use `if` without braces? `if (variableName in variables) return variables[variableName].toString()`

Reply:
Wow! Slowly coming back from covid, I see that there were some nice heated discussions around here 🙃 
We've been discussing the use of the ternary operator (vs. if/else) a lot with @nop33, and I always answered the same thing: for me the most important thing is to make the code as readable as possible.
Everyone probably agrees on that: we shouldn't make it convoluted for the sake of reducing the line count.

I don't want to enter the whole `if else` debate now, but regarding the code above, it was easier for me to read and understand quickly @polarker's version (with my low energy covid brain). 

Now, I really think that we should stop wasting energy on this... As long as the code is functional, safe and readable, coding styles can be adjusted as we go. It's really not worth it to argue so much about it IMO ☮️